src/coreclr/src/binder/inc/bindertracing.h

```diff
@@ -46,9 +53,128 @@ namespace BinderTracing
         bool m_checkedIgnoreBind;
         bool m_ignoreBind;
 
-        ReleaseHolder<PEAssembly> m_resultAssembly;
+        PEAssembly *m_resultAssembly;
         bool m_cached;
     };
+
+
+    // An object of this class manages firing events for all the stages during a binder resolving
+    // attempt operation.  It has minimal cost if tracing for this event is disabled.
+    //
+    // This class should be declared in the stack.  As information is determined by each stage
+    // (e.g. an AssemblySpec is initialized), the appropriate Set*() method should be called.  All
+    // pointers held by an object of this class must either be a nullptr, or point to a valid
+    // object during the time it is in scope.
+    //
+    // As the resolution progresses to different stages, the GoToStage() method should be called.
+    // Calling it will fire an event for the previous stage with whatever context the class had
+    // at that point; it is assumed that if GoToStage() is called, the previous stage failed
+    // (the HRESULT is read by the dtor to assess success).
+    //
+    // It holds a reference to a HRESULT (that must be live during the lifetime of this object),
+    // which is used to determine the success or failure of a stage either at the moment this
+    // class is destructed (e.g. last stage), or when moving from one stage to another.  (This
+    // is especially useful if the HRESULT is captured from an exception handler.)
+    class ResolutionAttemptedOperation
+    {
+    public:
+        // This must match the ResolutionAttemptedStage value map in ClrEtwAll.man
+        enum class Stage : uint16_t
+        {
+            FindInLoadContext = 0,
+            AssemblyLoadContextLoad = 1,
+            PlatformAssemblies = 2,
+            DefaultAssemblyLoadContextFallback = 3,
+            ResolveSatelliteAssembly = 4,
+            AssemblyLoadContextResolvingEvent = 5,
+            AppDomainAssemblyResolveEvent = 6,
+            NotYetStarted = 0xffff, // Used as flag to not fire event; not present in value map
+        };
+
+        struct AttemptInfo
```

Comment:
I don't see a need for this structure to be honest. We don't seem to expose it outside of the operation class and it's only used as a field which is always populated in the ctor. I would simply add the necessary fields of it onto the parent class.